// @roots/bud-sass/lib/extension.js
import { __decorate, __metadata } from "tslib";
import { Extension } from '@roots/bud-framework/extension';
import { bind, dependsOn, dependsOnOptional, expose, label, } from '@roots/bud-framework/extension/decorators';
import { BudSassOptions } from './options.js';
/**
 * Sass configuration
 */
export let BudSass = class BudSass extends BudSassOptions {
    /**
     * {@link Extension.boot}
     */
    async boot({ build, postcss }) {
        postcss?.setSyntax(`postcss-scss`);
        build.rules.sass.setUse(() => [
            build.items[`precss`],
            build.items[`css`],
            build.items[`postcss`],
            build.items[`resolve-url`],
            build.items[`sass`],
        ].filter(Boolean));
    }
    /**
     * {@link Extension.register}
     */
    async register({ build, hooks }) {
        /** Source loader */
        const loader = await this.resolve(`sass-loader`);
        if (!loader)
            return this.logger.error(`sass-loader not found`);
        /** Source sass implementation */
        const implementation = await this.import(`sass`, import.meta.url, {
            raw: true,
        });
        if (!implementation) {
            return this.logger.error(`sass not found`);
        }
        if (implementation.info) {
            this.setImplementation(implementation);
            this.logger.info(`sass implementation set (import * as sass)`, this.getImplementation()?.info);
        }
        else if (!implementation.info && implementation.default) {
            this.setImplementation(implementation.default);
            this.logger.info(`sass implementation set (import default)`, this.getImplementation()?.info);
        }
        else {
            this.logger.warning(`sass implementation not explicitly resolvable. falling back on sass loader default implementation.`);
        }
        /** Set loader alias */
        hooks.on(`build.resolveLoader.alias`, (aliases = {}) => ({
            ...aliases,
            [`sass-loader`]: loader,
        }));
        /** Resolve .scss and .sass extensions */
        hooks.on(`build.resolve.extensions`, ext => ext.add(`.scss`).add(`.sass`));
        /** Setup rule */
        build
            .setLoader(`sass`, `sass-loader`)
            .setItem(`sass`, {
            ident: `sass`,
            loader: `sass`,
            options: () => this.options,
        })
            .setRule(`sass`, {
            include: [({ path }) => path(`@src`)],
            test: ({ hooks }) => hooks.filter(`pattern.sass`),
        });
    }
};
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], BudSass.prototype, "boot", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Promise)
], BudSass.prototype, "register", null);
BudSass = __decorate([
    label(`@roots/bud-sass`),
    dependsOn([`@roots/bud-sass/resolve-url`]),
    dependsOnOptional([`@roots/bud-postcss`]),
    expose(`sass`)
], BudSass);
export {};
