// @roots/bud-postcss/lib/options.js
import { __decorate } from "tslib";
import { Extension, } from '@roots/bud-framework/extension';
import { options } from '@roots/bud-framework/extension/decorators';
let BudPostCssOptionsApi = class BudPostCssOptionsApi extends Extension {
    /**
     * PostCss config
     */
    config;
    /**
     * Get postcss config
     */
    getConfig;
    /**
     * Get PostCss plugin order
     */
    getOrder;
    /**
     * Get PostCss parser
     */
    getParser;
    /**
     * Get PostCss plugins
     */
    getPlugins;
    /**
     * get postcssOptions
     */
    getPostcssOptions;
    /**
     * get sourceMap option
     */
    getSourceMap;
    /**
     * Get PostCss syntax
     */
    getSyntax;
    /**
     * PostCss plugin order
     */
    order;
    /**
     * Parser
     */
    parser;
    /**
     * PostCss plugins
     */
    plugins;
    /**
     * postcssOptions
     */
    postcssOptions;
    /**
     * Set postcss config
     */
    setConfig;
    /**
     * Set PostCss plugin order
     */
    setOrder;
    /**
     * Set PostCss parser
     */
    setParser;
    /**
     * Set PostCss plugins
     */
    setPlugins;
    /**
     * set postcssOptions
     */
    setPostcssOptions;
    /**
     * set sourceMap option
     */
    setSourceMap;
    /**
     * Set PostCss syntax
     */
    setSyntax;
    /**
     * sourceMap option
     */
    sourceMap;
    /**
     * Syntax
     */
    syntax;
};
BudPostCssOptionsApi = __decorate([
    options({
        config: false,
        order: [],
        parser: undefined,
        plugins: {},
        postcssOptions: {},
        sourceMap: false,
        syntax: undefined,
    })
], BudPostCssOptionsApi);
export { BudPostCssOptionsApi, };
